language/types_values_variables.md

```diff
@@ -2185,6 +2188,52 @@ There is no automatic unwrapping of sensitive values. As a consequence it is not
 
 Sensitive values can be used in resources. It is not allowed to use a sensitive value as a resource title.
 
+### Deferred Type
+
+The type `Deferred` describes a function call for the purpose of resolving it in the future. It is a data object with
+the attributes `String name` and `Array[Any] attributes`.
+
+* When a deferred value is placed in a catalog an agent applying the catalog will resolve and replace
+  all such values with the result of the resolution prior to applying the catalog.
+* The order in which a catalog resolves the deferred values is undefined except when they appear as arguments
+  to another deferred, in which case they are resolved in the given order from left to right and depth first.
+* If a deferred is wrapped in a `Sensitive`, or if the resolution of the deferred results in a `Sensitive` then
+  the resulting value is handled just as if the catalog entry had been a sensitive value.
+* It is not possible to make a deferred call to a function that requires a lambda.
+
+#### Deferred.new
+
+Creates a new instance of the `Deferred` type. It accepts the name of the function to later call, and an array of
+arguments to use when making the call. The arguments are evaluated at the time the deferred value is created, but
+any deferred values among the arguments will be resolved as part of the resolution process (when actually making the call).
+
+    $x = Deferred('some_function', ['argument1', 42])
+
+Will, when the deferred is resolved result in the call:
+
+    some_function('argument1', 42)
```

Comment:
and presumably these would be ruby functions? or pp functions too? Or something else entirely (would they have scope and facts?)

Reply:
In first release only ruby functions. They are called in a global scope. `$facts` is set with the facts when deferred is resolved on the agent side.
(If resolved while compiling - if you call the deferred, it has everything the compiler has).

Reply:
nice, ok thanks

Reply:
PR updated with more details